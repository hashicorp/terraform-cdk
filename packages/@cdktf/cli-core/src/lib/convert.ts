/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import * as hcl2cdk from "@cdktf/hcl2cdk";
import { CdktfConfig } from "./cdktf-config";
import {
  ConstructsMakerProviderTarget,
  TerraformProviderConstraint,
  readSchema,
} from "@cdktf/provider-generator";
import { Errors, LANGUAGES } from "@cdktf/commons";

export async function convertConfigurationFile(configuration: string) {
  const cfg = CdktfConfig.read(process.cwd()); // TODO: make this the project directory instead of cwd
  const targets = cfg.terraformProviders.map((constraint) =>
    ConstructsMakerProviderTarget.from(
      new TerraformProviderConstraint(constraint),
      LANGUAGES[0]
    )
  );
  const { providerSchema } = await readSchema(targets);
  if (!providerSchema) {
    throw Errors.Internal("Could not fetch provider schema");
  }

  const { all } = await hcl2cdk.convert(configuration, {
    providerSchema,
    language: cfg.language,
    codeContainer: "constructs.Construct",
  });

  return all;
}
