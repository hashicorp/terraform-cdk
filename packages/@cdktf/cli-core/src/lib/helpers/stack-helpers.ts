// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0

import { ensureAllSettledBeforeThrowing, Errors, logger } from "@cdktf/commons";
import minimatch from "minimatch";
import { CdktfStack } from "../cdktf-stack";
import { SynthesizedStack } from "../synth-stack";

export function getSingleStack(
  stacks: SynthesizedStack[],
  stackName?: string,
  targetAction?: string,
) {
  if (!stacks) {
    throw Errors.Internal(
      "Trying to access a stack before it has been synthesized",
    );
  }

  if (stackName) {
    const stack = stacks.find((s) => s.name === stackName);
    if (!stack) {
      throw Errors.Usage("Could not find stack: " + stackName);
    }

    return stack;
  }

  if (stacks.length === 1) {
    return stacks[0];
  }

  throw Errors.Usage(
    `Found more than one stack, please specify a target stack. Run cdktf ${
      targetAction || "<verb>"
    } <stack> with one of these stacks: ${stacks
      .map((s) => s.name)
      .join(", ")} `,
  );
}

export function getMultipleStacks(
  stacks: SynthesizedStack[],
  patterns?: string[],
  targetAction?: string,
) {
  if (!patterns || !patterns.length) {
    if (stacks.length === 1) {
      return [stacks[0]];
    }
    throw Errors.Usage(
      `Found more than one stack, please specify a target stack. Run cdktf ${
        targetAction || "<verb>"
      } <stack> with one of these stacks: ${stacks
        .map((s) => s.name)
        .join(", ")} `,
    );
  }

  return patterns.flatMap((pattern) => {
    const matchingStacks = stacks.filter((stack) =>
      minimatch(stack.name, pattern),
    );

    if (matchingStacks.length === 0) {
      throw Errors.Usage(`Could not find stack for pattern '${pattern}'`);
    }

    return matchingStacks;
  });
}

// Returns the first stack that has no unmet dependencies
// An unmet dependency is a dependency that has not been deployed yet
// If there is no unfinished stack, returns undefined
// If there is no stack ready to be worked on, it returns a promise that will resolve as soon as there is a follow-up stack available
export async function getStackWithNoUnmetDependencies(
  stackExecutors: CdktfStack[],
): Promise<CdktfStack | undefined> {
  logger.debug("Checking for stacks with unmet dependencies");
  logger.debug("stack executors:", stackExecutors);
  const pendingStacks = stackExecutors.filter((executor) => executor.isPending);
  logger.debug("pending stacks:", stackExecutors);

  if (pendingStacks.length === 0) {
    return undefined;
  }

  const currentlyReadyStack = pendingStacks.find((executor) =>
    executor.stack.dependencies.every(
      (dependency) =>
        stackExecutors.find((executor) => executor.stack.name === dependency)
          ?.currentState === "done",
    ),
  );

  if (currentlyReadyStack) {
    logger.debug("Found a stack with no unmet dependencies");
    return currentlyReadyStack;
  }

  const stackExecutionPromises = stackExecutors
    .filter((ex) => ex.currentWorkPromise)
    .map((ex) => ex.currentWorkPromise);

  logger.debug(
    `${stackExecutionPromises.length} stacks are currently busy, waiting for one to finish`,
  );

  if (!stackExecutionPromises.length) {
    return undefined;
  }

  await ensureAllSettledBeforeThrowing(
    Promise.race(stackExecutionPromises),
    stackExecutionPromises,
  );

  return await getStackWithNoUnmetDependencies(stackExecutors);
}

function findAllDependantStacks(
  stackExecutors: CdktfStack[],
  stackName: string,
): CdktfStack[] {
  return stackExecutors.filter((innerExecutor) =>
    innerExecutor.stack.dependencies.includes(stackName),
  );
}

export function findAllNestedDependantStacks(
  stackExecutors: CdktfStack[],
  stackName: string,
  knownDependantStackNames: Set<string> = new Set(),
): CdktfStack[] {
  const dependantStacks = findAllDependantStacks(stackExecutors, stackName);
  dependantStacks.forEach((stack) => {
    if (knownDependantStackNames.has(stack.stack.name)) {
      return;
    }

    knownDependantStackNames.add(stack.stack.name);
    findAllNestedDependantStacks(
      stackExecutors,
      stack.stack.name,
      knownDependantStackNames,
    );
  });

  return stackExecutors.filter((executor) =>
    knownDependantStackNames.has(executor.stack.name),
  );
}

// Returns the first stack that has no dependents that need to be destroyed first
export async function getStackWithNoUnmetDependants(
  stackExecutors: CdktfStack[],
): Promise<CdktfStack | undefined> {
  logger.debug("Checking for stacks with unmet dependants");
  logger.debug("stack executors:", stackExecutors);
  const pendingStacks = stackExecutors.filter((executor) => executor.isPending);
  logger.debug("pending stacks:", stackExecutors);

  if (pendingStacks.length === 0) {
    return undefined;
  }

  const currentlyReadyStack = pendingStacks.find((executor) => {
    const dependantStacks = findAllDependantStacks(
      stackExecutors,
      executor.stack.name,
    );
    return dependantStacks.every((stack) => stack.currentState === "done");
  });

  if (currentlyReadyStack) {
    logger.debug("Found a stack with no unmet dependants");
    return currentlyReadyStack;
  }
  const stackExecutionPromises = stackExecutors
    .filter((ex) => ex.currentWorkPromise)
    .map((ex) => ex.currentWorkPromise);

  logger.debug(
    `${stackExecutionPromises.length} stacks are currently busy, waiting for one to finish`,
  );
  if (!stackExecutionPromises.length) {
    return undefined;
  }

  await Promise.race(stackExecutionPromises);
  return await getStackWithNoUnmetDependants(stackExecutors);
}

// Throws an error if there is a dependant stack that is not included
export function checkIfAllDependantsAreIncluded(
  stacksToRun: SynthesizedStack[],
  allStacks: SynthesizedStack[],
) {
  const allDependants = new Set<string>();
  stacksToRun
    .map((stack) =>
      allStacks.filter((s) => s.dependencies.includes(stack.name)),
    )
    .flat()
    .forEach((dependant) => allDependants.add(dependant.name));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependants = [...allDependants].filter(
    (dependant) => !stackNames.includes(dependant),
  );

  if (missingDependants.length > 0) {
    throw Errors.Usage(
      `The following dependant stacks are not included in the stacks to run: ${missingDependants.join(
        ", ",
      )}. Either add them or add the --ignore-missing-stack-dependencies flag.`,
    );
  }
}

/** 
  Throws an error if there is a dependency that is not included
  Cycles are detected on dependency creation at synthesis time
  Running this prevents us from being in a situation where we 
  have to wait for a stack to be deployed that is not included to be run
*/
export function checkIfAllDependenciesAreIncluded(
  stacksToRun: SynthesizedStack[],
) {
  const allDependencies = new Set<string>();
  stacksToRun
    .map((stack) => stack.dependencies)
    .flat()
    .forEach((dependency) => allDependencies.add(dependency));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependencies = [...allDependencies].filter(
    (dependency) => !stackNames.includes(dependency),
  );

  if (missingDependencies.length > 0) {
    throw Errors.Usage(
      `The following dependencies are not included in the stacks to run: ${missingDependencies.join(
        ", ",
      )}. Either add them or add the --ignore-missing-stack-dependencies flag.`,
    );
  }
}
