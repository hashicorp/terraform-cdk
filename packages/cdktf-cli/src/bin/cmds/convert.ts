// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import yargs from "yargs";
import { Errors } from "@cdktf/commons";
import { BaseCommand } from "./helper/base-command";
import { requireHandlers } from "./helper/utilities";
import * as path from "path";
import * as fs from "fs-extra";

function readCdktfJson(cwd = process.cwd()): { language: string } | undefined {
  try {
    const cdktfJsonPath = path.join(cwd, "cdktf.json");
    return fs.readJsonSync(cdktfJsonPath);
  } catch (e) {
    return undefined;
  }
}

class Command extends BaseCommand {
  public readonly command = "convert";
  public readonly describe =
    "Converts a single file of HCL configuration to CDK for Terraform. Takes the file to be converted on stdin.";

  public readonly builder = (args: yargs.Argv) =>
    args
      .example(
        "cat main.tf | cdktf convert  --provider integrations/github",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content and prints it"
      )
      .example(
        "cat main.tf | cdktf convert --provider hashicorp/aws > imported.ts",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts"
      )
      .example(
        "cat main.tf | cdktf convert --provider 'hashicorp/aws@ ~>3.62.0' 'integrations/github@ ~>4.16.0' --language python > imported.py",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.py"
      )
      .option("language", {
        choices: ["typescript", "python", "csharp", "java", "go"],
        default: readCdktfJson()?.language || "typescript",
      })
      .option("provider", {
        describe:
          "The conversion needs to know which providers are used in addition to the ones in your cdktf.json file. We search for a cdktf.json below your current working directory.",
        type: "array",
        default: [],
      })
      .option("stack", {
        describe: "Wrap the generated code within a stack class",
        type: "boolean",
        default: false,
      })
      .showHelpOnFail(true);

  public async handleCommand(argv: any) {
    Errors.setScope("convert");
    // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
    const api = requireHandlers();
    if (argv.withProject !== undefined) {
      if (argv.language === "typescript") {
        argv.withProject = false;
      } else {
        argv.withProject = true;
      }
    }
    await api.convert(argv);
  }
}

module.exports = new Command();
