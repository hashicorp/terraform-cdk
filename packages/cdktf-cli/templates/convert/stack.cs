using System;
using Constructs;
using HashiCorp.Cdktf;
/* {{ imports }} */

namespace MyCompany.MyApp
{
    class MyStack : TerraformStack
    {
        public MyStack(Construct scope, string id) : base(scope, id)
        {
            /* {{ code }} */
        }
    }
}