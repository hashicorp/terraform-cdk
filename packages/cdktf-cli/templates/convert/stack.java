package com.mycompany.app;

import software.constructs.Construct;
import com.hashicorp.cdktf.TerraformStack;
/* {{ imports }} */

public class MyStack extends TerraformStack
{
    public MyStack(final Construct scope, final String id) {
        super(scope, id);

        /* {{ code }} */
    }
}