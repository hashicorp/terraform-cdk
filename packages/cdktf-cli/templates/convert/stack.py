#!/usr/bin/env python
# Copyright (c) HashiCorp, Inc.
# SPDX-License-Identifier: MPL-2.0

from constructs import Construct
from cdktf import TerraformStack
# {{ imports }}


class MyStack(TerraformStack):
    def __init__(self, scope: Construct, id: str):
        super().__init__(scope, id)

        # {{ code }}
