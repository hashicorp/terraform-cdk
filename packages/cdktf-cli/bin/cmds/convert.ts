import yargs from "yargs";
import { Errors } from "../../lib/errors";
import { requireHandlers } from "./helper/utilities";

class Command implements yargs.CommandModule {
  public readonly command = "convert [OPTIONS]";
  public readonly describe =
    "Converts a single file of HCL configuration to CDK for Terraform. Takes the file to be converted on stdin.";

  public readonly builder = (args: yargs.Argv) =>
    args
      .example(
        "cat main.tf | cdktf convert  --provider integrations/github",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content and prints it"
      )
      .example(
        "cat main.tf | cdktf convert --provider hashicorp/aws > imported.ts",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts"
      )
      .example(
        "cat main.tf | cdktf convert --provider 'hashicorp/aws@ ~>3.62.0' 'integrations/github@ ~>4.16.0' --language python > imported.py",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts"
      )
      .option("language", {
        choices: ["typescript", "python", "csharp", "java"],
        default: "typescript",
      })
      .option("provider", {
        describe:
          "The conversion needs to know which providers are used in addition to the ones in your cdktf.json file. We search for a cdktf.json below your current working directory.",
        type: "array",
        default: [],
      })
      .showHelpOnFail(true);

  public async handler(argv: any) {
    Errors.setScope("convert");
    // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
    const api = requireHandlers();

    await api.convert(argv);
  }
}

module.exports = new Command();
