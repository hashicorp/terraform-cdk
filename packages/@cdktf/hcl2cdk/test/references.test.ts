// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { testCase, Synth, binding, Snapshot } from "./helpers/convert";

describe("references", () => {
  testCase.test(
    "resource references",
    `
        provider "aws" {
          region                      = "us-east-1"
        }
        resource "aws_kms_key" "examplekms" {
          description             = "KMS key 1"
          deletion_window_in_days = 7
        }

        resource "aws_s3_bucket" "examplebucket" {
          bucket = "examplebuckettftest"
          acl    = "private"
        }

        resource "aws_s3_bucket_object" "examplebucket_object" {
          key        = "someobject"
          bucket     = aws_s3_bucket.examplebucket.id
          source     = "index.html"
          kms_key_id = aws_kms_key.examplekms.arn
        }`,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    {
      resources: ["aws_s3_bucket", "aws_s3_bucket_object", "aws_kms_key"],
    }
  );

  testCase.test(
    "locals references",
    `
        provider "aws" {
          region                      = "us-east-1"
        }

        locals {
          bucket_name = "foo"
        }

        resource "aws_s3_bucket" "examplebucket" {
          bucket = local.bucket_name
          acl    = "private"
        }`,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    {
      resources: ["aws_s3_bucket"],
    }
  );

  testCase.test(
    "variable references",
    `
        provider "aws" {
          region                      = "us-east-1"
        }
        variable "bucket_name" {
          type    = string
          default = "demo"
        }

        resource "aws_s3_bucket" "examplebucket" {
          bucket = var.bucket_name
          acl    = "private"
        }`,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    {
      resources: ["aws_s3_bucket"],
    }
  );

  testCase.test(
    "data references",
    `
        provider "aws" {
          region                      = "us-east-1"
        }
        variable "bucket_name" {
          type    = string
          default = "demo"
        }

        data "aws_s3_bucket" "examplebucket" {
          bucket = var.bucket_name
        }

        resource "aws_s3_bucket_object" "examplebucket_object" {
          key        = "someobject"
          bucket     = data.aws_s3_bucket.examplebucket.arn
          source     = "index.html"
        }`,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    {
      resources: ["aws_s3_bucket", "aws_s3_bucket_object"],
    }
  );

  testCase.test(
    "double references",
    `
        provider "aws" {
          region                      = "us-east-1"
        }
        variable "bucket_name" {
          type    = string
          default = "demo"
        }

        resource "aws_s3_bucket" "examplebucket" {
          bucket = var.bucket_name
          acl    = "private"
          tags = {
            tag-key = var.bucket_name
          }
        }`,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    {
      resources: ["aws_s3_bucket"],
    }
  );

  testCase.test(
    "resource references with HCL functions",
    `
      provider "aws" {
        region                      = "us-east-1"
      }
      resource "aws_kms_key" "examplekms" {
        description             = "KMS key 1"
        deletion_window_in_days = 7
      }

      resource "aws_s3_bucket" "examplebucket" {
        count  = 2
        bucket = "examplebuckettftest-\${count.index}"
        acl    = "private"
      }

      resource "aws_s3_bucket_object" "examplebucket_object" {
        key        = "someobject"
        bucket     = element(aws_s3_bucket.examplebucket, 0).id
        source     = "index.html"
        kms_key_id = aws_kms_key.examplekms.arn
      }`,
    [binding.aws],
    Snapshot.yes,
    Synth.no_cant_resolve_construct,
    {
      resources: ["aws_s3_bucket", "aws_s3_bucket_object", "aws_kms_key"],
    }
  );

  testCase.test(
    "provider with var reference",
    `
    variable "domain" {
      description = "A domain"
    }
    provider "auth0" {
      domain        = var.domain
      client_id     = "client_id"
      client_secret = "client_secret"
    }
    `,
    [binding.auth0],
    Snapshot.yes,
    Synth.yes,
    { resources: [] }
  );

  testCase.test(
    "data local_file",
    `
    provider "local" {}
    data "local_file" "_01_please_verify" {
      filename = "./email_templates/01_please_verify/template.html"
    }
    `,
    [binding.local],
    Snapshot.yes,
    Synth.yes,
    {
      dataSources: ["local_file"],
    }
  );

  testCase.test(
    "aliased duplicate provider with var reference",
    `
    variable "domain" {
      description = "A domain"
    }
    variable "client_id" {
      description = "A client_id"
    }
    variable "client_secret" {
      description = "A client_secret"
    }
    provider "auth0" {
      domain        = var.domain
      client_id     = var.client_id
      client_secret = var.client_secret
    }
    provider "auth0" {
      alias         = "private_auth0"
      domain        = var.domain
      client_id     = var.client_id
      client_secret = var.client_secret
      debug         = true
    }
    `,
    [binding.auth0],
    Snapshot.yes,
    Synth.yes,
    { resources: [] }
  );

  testCase.test(
    "variables with maps need to use accessor syntax",
    `
      provider "aws" {
        region = "us-east-1"
      }

      variable "default_tags" {
        type        = map(string)
        description = "Map of default tags to apply to resources"
        default = {
          project = "Learning Live with AWS & HashiCorp"
        }
      }

      resource "aws_eip" "nat" {
        vpc = true
        tags = {
          "Name" = "\${var.default_tags.project}-nat-eip"
        }
      }
      `,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    { resources: ["aws_eip"] }
  );

  testCase.test(
    "resources with splat expressions should work",
    `
      variable "key_value_pairs" {
        type        = list(object({
          foo = string
          bar = string
        }))
        description = "List of key value pairs"
        default = [{
          foo = "foo"
          bar = "not food",
        }, {
          foo = "bar"
          bar = "not beer",
        }]
      }

      output "values" {
        value = var.key_value_pairs[*].bar
      }
      `,
    [],
    Snapshot.yes,
    Synth.yes,
    { resources: ["aws_eip"] }
  );
});
