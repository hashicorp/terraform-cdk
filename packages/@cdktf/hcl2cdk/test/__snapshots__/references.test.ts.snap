// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`references aliased duplicate provider with var reference snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { TerraformVariable, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { Auth0Provider } from "./.gen/providers/auth0/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    const clientId = new TerraformVariable(this, "client_id", {
      description: "A client_id",
    });
    const clientSecret = new TerraformVariable(this, "client_secret", {
      description: "A client_secret",
    });
    const domain = new TerraformVariable(this, "domain", {
      description: "A domain",
    });
    new Auth0Provider(this, "auth0", {
      clientId: clientId.stringValue,
      clientSecret: clientSecret.stringValue,
      domain: domain.stringValue,
    });
    new Auth0Provider(this, "auth0_4", {
      alias: "private_auth0",
      clientId: clientId.stringValue,
      clientSecret: clientSecret.stringValue,
      debug: true,
      domain: domain.stringValue,
    });
  }
}
"
`;

exports[`references data local_file snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { DataLocalFile } from "./.gen/providers/local/data-local-file";
import { LocalProvider } from "./.gen/providers/local/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    new LocalProvider(this, "local", {});
    new DataLocalFile(this, "_01_please_verify", {
      filename: "./email_templates/01_please_verify/template.html",
    });
  }
}
"
`;

exports[`references data references snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { VariableType, TerraformVariable, Token, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { DataAwsS3Bucket } from "./.gen/providers/aws/data-aws-s3-bucket";
import { AwsProvider } from "./.gen/providers/aws/provider";
import { S3BucketObject } from "./.gen/providers/aws/s3-bucket-object";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const bucketName = new TerraformVariable(this, "bucket_name", {
      default: "demo",
      type: VariableType.STRING,
    });
    const examplebucket = new DataAwsS3Bucket(this, "examplebucket", {
      bucket: bucketName.stringValue,
    });
    new S3BucketObject(this, "examplebucket_object", {
      bucket: Token.asString(examplebucket.arn),
      key: "someobject",
      source: "index.html",
    });
  }
}
"
`;

exports[`references double references snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { VariableType, TerraformVariable, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { AwsProvider } from "./.gen/providers/aws/provider";
import { S3Bucket } from "./.gen/providers/aws/s3-bucket";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const bucketName = new TerraformVariable(this, "bucket_name", {
      default: "demo",
      type: VariableType.STRING,
    });
    new S3Bucket(this, "examplebucket", {
      acl: "private",
      bucket: bucketName.stringValue,
      tags: {
        "tag-key": bucketName.stringValue,
      },
    });
  }
}
"
`;

exports[`references locals references snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { AwsProvider } from "./.gen/providers/aws/provider";
import { S3Bucket } from "./.gen/providers/aws/s3-bucket";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const bucketName = "foo";
    new S3Bucket(this, "examplebucket", {
      acl: "private",
      bucket: bucketName,
    });
  }
}
"
`;

exports[`references provider with var reference snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { TerraformVariable, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { Auth0Provider } from "./.gen/providers/auth0/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    const domain = new TerraformVariable(this, "domain", {
      description: "A domain",
    });
    new Auth0Provider(this, "auth0", {
      clientId: "client_id",
      clientSecret: "client_secret",
      domain: domain.stringValue,
    });
  }
}
"
`;

exports[`references resource references snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { KmsKey } from "./.gen/providers/aws/kms-key";
import { AwsProvider } from "./.gen/providers/aws/provider";
import { S3Bucket } from "./.gen/providers/aws/s3-bucket";
import { S3BucketObject } from "./.gen/providers/aws/s3-bucket-object";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const examplekms = new KmsKey(this, "examplekms", {
      deletionWindowInDays: 7,
      description: "KMS key 1",
    });
    const examplebucket = new S3Bucket(this, "examplebucket", {
      acl: "private",
      bucket: "examplebuckettftest",
    });
    new S3BucketObject(this, "examplebucket_object", {
      bucket: examplebucket.id,
      key: "someobject",
      kmsKeyId: examplekms.arn,
      source: "index.html",
    });
  }
}
"
`;

exports[`references resource references with HCL functions snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { Token, TerraformCount, Fn, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { KmsKey } from "./.gen/providers/aws/kms-key";
import { AwsProvider } from "./.gen/providers/aws/provider";
import { S3Bucket } from "./.gen/providers/aws/s3-bucket";
import { S3BucketObject } from "./.gen/providers/aws/s3-bucket-object";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const examplekms = new KmsKey(this, "examplekms", {
      deletionWindowInDays: 7,
      description: "KMS key 1",
    });
    /*In most cases loops should be handled in the programming language context and 
    not inside of the Terraform context. If you are looping over something external, e.g. a variable or a file input
    you should consider using a for loop. If you are looping over something only known to Terraform, e.g. a result of a data source
    you need to keep this like it is.*/
    const examplebucketCount = TerraformCount.of(Token.asNumber("2"));
    const examplebucket = new S3Bucket(this, "examplebucket", {
      acl: "private",
      bucket: "examplebuckettftest-\${" + examplebucketCount.index + "}",
      count: examplebucketCount,
    });
    new S3BucketObject(this, "examplebucket_object", {
      bucket: Token.asString(
        Fn.lookupNested(Fn.element(examplebucket, 0), ["id"])
      ),
      key: "someobject",
      kmsKeyId: examplekms.arn,
      source: "index.html",
    });
  }
}
"
`;

exports[`references resources with splat expressions should work snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  TerraformOutput,
  Fn,
  TerraformStack,
} from "cdktf";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    const keyValuePairs = new TerraformVariable(this, "key_value_pairs", {
      default: [
        {
          bar: "not food",
          foo: "foo",
        },
        {
          bar: "not beer",
          foo: "bar",
        },
      ],
      description: "List of key value pairs",
      type: VariableType.list(
        VariableType.object({
          bar: VariableType.STRING,
          foo: VariableType.STRING,
        })
      ),
    });
    new TerraformOutput(this, "values", {
      value: Fn.lookupNested(keyValuePairs.value, ["*", "bar"]),
    });
  }
}
"
`;

exports[`references variable references snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { VariableType, TerraformVariable, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { AwsProvider } from "./.gen/providers/aws/provider";
import { S3Bucket } from "./.gen/providers/aws/s3-bucket";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const bucketName = new TerraformVariable(this, "bucket_name", {
      default: "demo",
      type: VariableType.STRING,
    });
    new S3Bucket(this, "examplebucket", {
      acl: "private",
      bucket: bucketName.stringValue,
    });
  }
}
"
`;

exports[`references variables with maps need to use accessor syntax snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  Fn,
  Token,
  TerraformStack,
} from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { Eip } from "./.gen/providers/aws/eip";
import { AwsProvider } from "./.gen/providers/aws/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const defaultTags = new TerraformVariable(this, "default_tags", {
      default: [
        {
          project: "Learning Live with AWS & HashiCorp",
        },
      ],
      description: "Map of default tags to apply to resources",
      type: VariableType.map(VariableType.STRING),
    });
    new Eip(this, "nat", {
      tags: {
        Name:
          Token.asString(Fn.lookupNested(defaultTags.value, ["project"])) +
          "-nat-eip",
      },
      vpc: true,
    });
  }
}
"
`;
